Add partial raw identifier support.

jonmeow opened the pull request:
I'm looking at this due to the conversation on #3341. Although diagnostics aren't where they should be, I thought it may help to start adding raw identifier support (which may also help show how I was thinking about this).

Note regarding the TODO on how to form the token, `GetTokenText` returns the `string_id`'s reference value for an `Identifier`. So to make `GetTokenText` work in a way that returns `r#foo` for a raw identifier, I think there are a few options:

1. Add additional data indicating the end of the identifier.
2. Add `RawIdentifier` as a token kind to indicate that it's raw and should be prefixed with `r#` (but also giving later stages one more token kind to handle)
3. Make the `string_id` correspond to `r#foo`, and have later stages add `foo` to the strings table whenever `r#foo` is encountered (with map lookups leading to deduplication).
4. Add `StringId::RawKeyword` special values for each keyword.
    - This would mean `self` prints as `self`, `r#self` prints as `r#self`, but `r#foo` is not a keyword so prints as `foo`.
    - This means keywords would need to be listed in a place `StringId` can depend on them, one way or the other (e.g., a `keywords.def` file in `base/` should work).
5. Say that it _is_ an `Identifier`, and if it's a keyword spelling, it must have been a raw identifier.
    - Same limitation as above: This would mean `self` prints as `self`, `r#self` prints as `r#self`, but `r#foo` is not a keyword so prints as `foo`.

I'm hoping to resolve this issue separately though. :)

jonmeow committed:
Add partial raw identifier support.

zygoloid commented on toolchain/lex/tokenized_buffer.cpp:
```diff
@@ -894,6 +900,42 @@ class [[clang::internal_linkage]] TokenizedBuffer::Lexer {
+    // Raw identifiers must look like `r#<valid identifier>`, otherwise it's an
+    // identifier starting with the 'r'.
+    // TODO: Need to add support for Unicode lexing.
+    if (position + 2 >= static_cast<ssize_t>(source_text.size()) ||
```
I wonder if it's worth putting an `LLVM_LIKELY` annotation on this `if` condition, or perhaps on the second part -- character 2 is unlikely to be `#`.

jonmeow replied:
Done.

zygoloid commented on toolchain/lex/tokenized_buffer.cpp:
```diff
@@ -894,6 +900,42 @@ class [[clang::internal_linkage]] TokenizedBuffer::Lexer {
+        {.kind = TokenKind::Identifier,
+         .token_line = current_line(),
+         .column = column,
+         .string_id = buffer_.value_stores_->strings().Add(identifier_text)});
```
In addition to the options presented in your comment, we have a spare sign bit in this field that we could use to track whether we have a raw identifier. But I'm fine with not tracking that in this PR.

zygoloid reviewed (approved):
Some thoughts, but looks good to me. It'd be good to check the lexing benchmarks, but given how they're set up this will only affect 1/52 of the cases. Maybe it's worth adding a benchmark for identifiers starting with `r` and raw identifiers?

jonmeow committed:
Comments

jonmeow committed:
bm

jonmeow commented:
I've added a couple focused benchmarks, but I'm expecting it should be a pretty low fraction of identifiers such that real-world performance impact is negligible (so long as the approach is reasonable, which for `r` it's just a couple extra comparisons). For benchmarks, the fact it's traversing a different code path (i.e., not just identifiers) may impact.

For benchmarks, I think @chandlerc's comparison technique would be needed to see if this is a real impact.

Here's a snippet of benchmarks, including the two "RawIdentifier"-related benchmarks I've added (AsRawIdentifiers is _only_ raw, RawIdentifierFocus is 50-50 split):

```
------------------------------------------------------------------------------------------------------
Benchmark                                            Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------------------
BM_ValidKeywords                               2790487 ns      2790425 ns          245 bytes_per_second=214.025M/s tokens_per_second=35.8368M/s
BM_ValidKeywordsAsRawIdentifiers               3368912 ns      3368880 ns          213 bytes_per_second=233.892M/s tokens_per_second=29.6835M/s
BM_RawIdentifierFocus                          6068617 ns      6067586 ns           95 bytes_per_second=257.886M/s tokens_per_second=16.481M/s
BM_ValidIdentifiers<1, 64, false>              6397688 ns      6396173 ns          102 bytes_per_second=230.265M/s tokens_per_second=15.6343M/s
BM_ValidIdentifiers<1, 1, true>                3563627 ns      3563433 ns          202 bytes_per_second=80.2881M/s tokens_per_second=28.0628M/s
BM_ValidIdentifiers<3, 5, true>                5155044 ns      5154598 ns          116 bytes_per_second=111.008M/s tokens_per_second=19.4002M/s
BM_ValidIdentifiers<3, 16, true>               4745501 ns      4745391 ns          121 bytes_per_second=231.111M/s tokens_per_second=21.0731M/s
BM_ValidIdentifiers<12, 64, true>              5344008 ns      5344022 ns          103 bytes_per_second=713.784M/s tokens_per_second=18.7125M/s
BM_ValidIdentifiers<16, 16, true>              5053913 ns      5053594 ns          106 bytes_per_second=339.682M/s tokens_per_second=19.7879M/s
BM_ValidIdentifiers<24, 24, true>              5095683 ns      5094748 ns          106 bytes_per_second=486.688M/s tokens_per_second=19.6281M/s
BM_ValidIdentifiers<32, 32, true>              5750146 ns      5749461 ns           94 bytes_per_second=563.964M/s tokens_per_second=17.3929M/s
BM_ValidIdentifiers<48, 48, true>              6042171 ns      6040988 ns           93 bytes_per_second=789.336M/s tokens_per_second=16.5536M/s
BM_ValidIdentifiers<64, 64, true>              7908147 ns      7907551 ns           76 bytes_per_second=795.98M/s tokens_per_second=12.6461M/s
BM_ValidIdentifiers<80, 80, true>              8130933 ns      8129182 ns           71 bytes_per_second=961.982M/s tokens_per_second=12.3014M/s
...
BM_RandomSource                                4918620 ns      4918372 ns          136 bytes_per_second=211.378M/s lines_per_second=6.45437M/s tokens_per_second=20.3319M/s
...
```

And at trunk:

```
------------------------------------------------------------------------------------------------------
Benchmark                                            Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------------------
BM_ValidKeywords                               2724324 ns      2724283 ns          254 bytes_per_second=219.221M/s tokens_per_second=36.7069M/s
BM_ValidIdentifiers<1, 64, false>              6186557 ns      6186579 ns          102 bytes_per_second=238.066M/s tokens_per_second=16.164M/s
BM_ValidIdentifiers<1, 1, true>                3378836 ns      3378755 ns          206 bytes_per_second=84.6766M/s tokens_per_second=29.5967M/s
BM_ValidIdentifiers<3, 5, true>                4992943 ns      4992668 ns          114 bytes_per_second=114.609M/s tokens_per_second=20.0294M/s
BM_ValidIdentifiers<3, 16, true>               4709036 ns      4709053 ns          117 bytes_per_second=232.895M/s tokens_per_second=21.2357M/s
BM_ValidIdentifiers<12, 64, true>              5296208 ns      5296218 ns          100 bytes_per_second=720.226M/s tokens_per_second=18.8814M/s
BM_ValidIdentifiers<16, 16, true>              4904774 ns      4904792 ns          110 bytes_per_second=349.987M/s tokens_per_second=20.3882M/s
BM_ValidIdentifiers<24, 24, true>              4786279 ns      4786199 ns          115 bytes_per_second=518.063M/s tokens_per_second=20.8934M/s
BM_ValidIdentifiers<32, 32, true>              5530390 ns      5530287 ns           98 bytes_per_second=586.315M/s tokens_per_second=18.0822M/s
BM_ValidIdentifiers<48, 48, true>              6207654 ns      6207676 ns           95 bytes_per_second=768.141M/s tokens_per_second=16.1091M/s
BM_ValidIdentifiers<64, 64, true>              7575676 ns      7575695 ns           77 bytes_per_second=830.848M/s tokens_per_second=13.2001M/s
BM_ValidIdentifiers<80, 80, true>              7782718 ns      7782247 ns           74 bytes_per_second=1004.87M/s tokens_per_second=12.8498M/s
...
BM_RandomSource                                4921775 ns      4921219 ns          142 bytes_per_second=211.256M/s lines_per_second=6.45064M/s tokens_per_second=20.3202M/s
...
```

zygoloid commented on toolchain/lex/tokenized_buffer_benchmark.cpp:
```diff
@@ -427,6 +427,66 @@ void BM_ValidKeywords(benchmark::State& state) {
+}
+BENCHMARK(BM_ValidKeywordsAsRawIdentifiers);
+
+// This benchmark does a 50-50 split of r-prefixed and r#-prefixed identifiers
```
Is this comparing `r#`-prefixed and `r`-prefixed identifiers, or `r#`-prefixed identifiers and general non-raw identifiers?

jonmeow replied:
`r`-prefixed, not general. Note, it looks like when you reviewed you were looking at the "Comments" commit but not the "bm" commit; the "Comments" commit did general.

https://github.com/carbon-language/carbon-lang/pull/3344/commits/2f5d94374e20d98c330bd074af4174bf214856ef
